src/event.rs

```diff
@@ -627,6 +634,28 @@ pub struct KeyboardInput {
     pub modifiers: ModifiersState,
 }
 
+/// Describes an event from input method.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum Ime {
+    /// Notifies when the IME was enabled.
+    Enabled,
+
+    /// Notifies when a new composing text should be set at the cursor position.
+    ///
+    /// The value represents a pair of the preedit string and the cursor begin position and end
+    /// position. When both indices are `None`, the cursor should be hidden.
+    ///
+    /// The cursor position is byte-wise indexed.
+    Preedit(String, Option<usize>, Option<usize>),
+
+    /// Notifies when text should be inserted into the editor widget.
+    Commit(String),
```

Comment:
I suspect it would make sense to merge this into `WindowEvent::ReceivedCharacter`, so clients only need to handle one event when they want to actually insert text into their buffer.

But this is probably a larger effort for lesser benefit, so I won't block on that.

Reply:
No, it won't. The point is that you should know that you've got a commit from the IME, so you can remove preedit area and insert text at once into the widget. I'm pretty sure everything will change with new keyboard API, but the IME text won't be attached to any keyboard event anyway, so it should be delivered via separate event, like commit, and it's more intuitive for the end user, since that's how every IME API works out there.